<?xml version="1.0" encoding="utf-8"?>
<launch>
  <arg name="bagfile"   />
  <arg name="nodelet_manager" default="darknet_ros_nodelet_manager" />
  <arg name="image_topic" default="/camera/left/image_raw" />
  <param name="/darknet_ros/image_view/enable_console_output" type="bool" value="false" />
  <param name="/darknet_ros/image_view/enable_opencv" type="bool" value="false" />


  <!-- play bag file -->
  <node pkg="rosbag" type="play" name="whuvid" args="-d 5 $(arg bagfile)" required="true">
    <remap from="/camera/left/image_raw" to="$(arg image_topic)" />
  </node>

  <!-- play usb camera
  <node pkg="cv_camera" type="cv_camera_node" name="cam" >
    <remap from="/cam/image_raw" to="$(arg image_topic)" />
  </node>
  -->

  <!-- run object detector  file  -->
  <arg name="bbox_topic" default="/camera/left/bounding_boxes" />
  <include file="$(find darknet_ros)/launch/darknet_ros_nodelet.launch" >

    <!-- ROS and network parameter files -->
    <arg name="ros_param_file"             default="$(find starmap)/launch/config/ros.yaml"/>
    <arg name="network_param_file" value="$(find darknet_ros)/config/yolov3.yaml" />
    <!-- arg name="launch_prefix" value="gdb -ex run -args"/ -->
    <!-- remap from="/camera/left/bounding_boxes" to="$(arg bbox_topic)" /-->
  </include>

  <!-- sort_ros.launch 
  -->
  <arg name="tracked_bbox_topic" default="/camera/left/tracked_bounding_boxes" />
  <arg name="visualization_topic" default="/camera/left/tracked_bbox_vis" />
  <node pkg="nodelet" type="nodelet" name="sort_ros"
        args="load sort_ros/SortRos $(arg nodelet_manager)"
        >
    <param name="image_topic" value="$(arg image_topic)" />
    <param name="bbox_topic" value="$(arg bbox_topic)" />
    <param name="tracked_bbox_topic" value="$(arg tracked_bbox_topic)" />
    <param name="visualization_topic" value="$(arg visualization_topic)" />
    <param name="valid_class_labels"  type="yaml"
           value="[car, truck, bus]"
           />
  </node>

  <!-- visualize sort ros image
  <node pkg="image_view" type="image_view" name="vis_sort_ros" >
    <remap from="image" to="$(arg visualization_topic)" />
  </node>
  -->

  <!-- starmap.launch 
  -->
  <arg name="keypoint_topic" default="/camera/left/keypoints" />
  <node pkg="nodelet" type="nodelet" name="starmap"
        args="load starmap/Starmap $(arg nodelet_manager)"
         >
    <param name="image_topic" value="$(arg image_topic)" />
    <param name="bbox_topic" value="$(arg tracked_bbox_topic)" />
    <param name="keypoint_topic" value="$(arg keypoint_topic)" />
    <param name="starmap_model_path" value="$(find starmap)/models/model_cpu-jit.pth" />
    <param name="gpu_id" value="0" type="int" />
    <param name="to_color_img_flag" type="bool" value="true" />
    <param name="visualize" value="false" type="bool" />
  </node>

  <!-- rosbag record
  -->
  <node pkg="rosbag" type="record" name="rosbag_record"
        args="$(eval
              arg('image_topic') + ' '
              + arg('tracked_bbox_topic') + ' '
              + arg('bbox_topic') + ' '
              + arg('keypoint_topic')
              + ' -O ' + arg('bagfile')[:-4] + '-bbox.bag'
              )"
        />
</launch>
